pilot/pkg/networking/core/v1alpha3/httproute.go

```diff
@@ -163,28 +164,40 @@ func (configgen *ConfigGeneratorImpl) buildSidecarOutboundHTTPRouteConfig(env *m
 	// Get list of virtual services bound to the mesh gateway
 	virtualHostWrappers := istio_route.BuildSidecarVirtualHostsFromConfigAndRegistry(node, push, nameToServiceMap, proxyLabels, virtualServices, listenerPort)
 	vHostPortMap := make(map[int][]route.VirtualHost)
-
+	var name string
+	uniques := make(map[string]struct{})
 	for _, virtualHostWrapper := range virtualHostWrappers {
 		// If none of the routes matched by source, skip this virtual host
 		if len(virtualHostWrapper.Routes) == 0 {
 			continue
 		}
-
 		virtualHosts := make([]route.VirtualHost, 0, len(virtualHostWrapper.VirtualServiceHosts)+len(virtualHostWrapper.Services))
 		for _, host := range virtualHostWrapper.VirtualServiceHosts {
-			virtualHosts = append(virtualHosts, route.VirtualHost{
-				Name:    fmt.Sprintf("%s:%d", host, virtualHostWrapper.Port),
-				Domains: []string{host, fmt.Sprintf("%s:%d", host, virtualHostWrapper.Port)},
-				Routes:  virtualHostWrapper.Routes,
-			})
+			name = fmt.Sprintf("%s:%d", host, virtualHostWrapper.Port)
+			if _, found := uniques[name]; !found {
```

Comment:
@costinm @rshriram, philosophical question here:  how hard should Pilot try to mirror Envoy's validation vs. delegating to Envoy and relaying the decision?

Reply:
Well we don’t have a feedback loop. That’s about the crux of it. So we try really hard to validate and automatically figure out stuff